### DISCUSSION TITLE
Performance Testing

### DISCUSSION THREAD

**Discussion by rajangdavis** [djipco/webmidi#322] [General]
I made an app to control my pedal board with WEBMIDI.js powering the MIDI commands and I have been experimenting with creating macro commands of MIDI CC, PC, and Sysex commands/messages.

I also have added some expression controls that stack CC commands across a few channels.

What I was curious about is how might it be possible to test for performance issues in either my codebase or with the WEBMIDI.js codebase in regards to MIDI.

When testing my app via Wireless Bluetooth MIDI and a Samsung tablet, the changes feel solid and _realtime enough_ meaning I haven't perceived any lag or performance issues.

My concern lies in maintaining the current baseline of how my app is performing currently. I also think having some sort of metrics around performance may help with WEBMIDI.js quality control and may help inform what kind of features may be added in the future.

Using a tool like [Morningstar's MIDI Monitor](https://www.midimonitor.com/), I am seeing these numbers in regards to a macro command that I am sending from a click event:

|Timestamp | Device | Type | Data | Channel |
| --- | --- | --- | --- | --- |
|23:13 50:876 | IAC Driver Bus 1 | System Exclusive | F0 00 20 10 00 01 01 26 03 70 0C 4D 7F 21 7F 00 15 58 7F 3F 00 7F 00 7F 00 00 70 0C 4D 7F 49 7F 00 15 58 00 3F 00 F7 |  | 
|23:13 50:876 | IAC Driver Bus 1 | System Exclusive | F0 00 20 10 00 01 02 26 02 7F 33 1E 2B 00 28 05 40 7F 05 7F 52 7F 00 00 00 78 7F 66 71 0F 00 28 05 40 7F 05 7F 52 F7 |  | 
|23:13 50:875 | IAC Driver Bus 1 | System Exclusive | F0 00 20 10 00 01 04 26 04 00 23 65 42 42 40 00 00 20 00 06 07 7F 00 5F 7F 0D 00 23 65 42 08 40 00 00 20 00 06 07 F7 |  | 
|23:13 50:875 | IAC Driver Bus 1 | System Exclusive | F0 00 20 10 00 01 03 26 07 3E 7F 7F 7F 3F 7F 00 54 5E 31 1E 54 7F 7F 3F 00 3F 00 7F 7F 21 60 35 00 54 12 00 66 2B F7 |  | 
|23:13 50:875 | IAC Driver Bus 1 | Control Change | CC#14 Value 127 | 4 | 
|23:13 50:874 | IAC Driver Bus 1 | Control Change | CC#14 Value 127 | 3 | 
|23:13 50:874 | IAC Driver Bus 1 | Control Change | CC#14 Value 127 | 2 | 
|23:13 50:874 | IAC Driver Bus 1 | Control Change | CC#14 Value 127 | 1 | 
|23:13 50:873 | IAC Driver Bus 1 | Control Change | CC#4 Value 0 | 4 | 
|23:13 50:873 | IAC Driver Bus 1 | Control Change | CC#4 Value 0 | 3 | 
|23:13 50:872 | IAC Driver Bus 1 | Control Change | CC#4 Value 0 | 2 | 
|23:13 50:872 | IAC Driver Bus 1 | Control Change | CC#4 Value 0 | 1 | 
|23:13 50:871 | IAC Driver Bus 1 | Control Change | CC#14 Value 0 | 6 | 
|23:13 50:869 | IAC Driver Bus 1 | Control Change | CC#14 Value 0 | 5 | 

Altogether, these numbers don't seem too bad in regards to a hacked together React app; however, I am clueless as far as what these numbers _should look like_, whether there is room for improvement, and if devices have different performance characteristics that I should be aware.

I have been a happy hacker with the WEBMIDI.js; hopefully should have something to share in the next year!

**Comment by djipco** [comment]
Thanks for posting the info. These numbers look good to me. 

I have performed tests, early in the development of WEBMIDI.js, to check whether the library was resulting in any performance degradation. My, admittedly limited, tests showed negligible degradation compared to native solutions. 

Having said that, I have not looked into performance since. It has not come up in the discussions so I never bothered. However, I do agree that it would be interesting to have some sort of procedure to assess performance.
